lib/sycamore/sycamore/transforms/extract_graph.py

```diff
@@ -232,75 +149,190 @@ async def gather_api_calls():
+        docset.plan = ResolveEntities.GroupEntities(docset.plan)
+        docset = ResolveEntities._resolve(docset)
+        return docset
+
+    class GroupEntities(Map):
+        def __init__(self, child: Node, **resource_args):
+            super().__init__(child, f=self.GroupDocumentNodes, **resource_args)
+
+        @staticmethod
+        def GroupDocumentNodes(doc: HierarchicalDocument) -> HierarchicalDocument:
+            if "EXTRACTED_NODES" in doc.data:
+                return doc
+            nodes: defaultdict[dict, Any] = defaultdict(dict)
+            nodes |= doc["properties"].get("nodes", {})
+            for section in doc.children:
+                if "nodes" not in section["properties"]:
+                    continue
+                for label, hashes in section["properties"]["nodes"].items():
+                    for hash, node in hashes.items():
+                        if nodes[label].get(hash, None) is None:
+                            nodes[label][hash] = node
+                        else:
+                            for rel_uuid, rel in node["relationships"].items():
+                                nodes[label][hash]["relationships"][rel_uuid] = rel
+                del section["properties"]["nodes"]
+            doc["properties"]["nodes"] = nodes
+            return doc
+
+    @staticmethod
+    def _resolve(docset: "DocSet") -> "DocSet":
+        from sycamore import Execution
+        from sycamore.reader import DocSetReader
+        from ray.data.aggregate import AggregateFn
+
+        reader = DocSetReader(docset.context)
+
+        # Get list[Document] representation of docset, trigger execute with take_all()
+        execution = Execution(docset.context, docset.plan)
+        dataset = execution.execute(docset.plan)
+        docs = dataset.take_all()
+        docs = [Document.deserialize(d["doc"]) for d in docs]
+
+        # Update docset and dataset to version after execute
+        docset = reader.document(docs)
+        execution = Execution(docset.context, docset.plan)
+        dataset = execution.execute(docset.plan)
+
+        def extract_nodes(row):
+            doc = Document.deserialize(row["doc"])
+            if isinstance(doc, MetadataDocument) or "nodes" not in doc["properties"]:
+                return {}
+            return doc["properties"]["nodes"]
+
+        def accumulate_row(nodes, row):
+            extracted = extract_nodes(row)
+            for key, hashes in extracted.items():
+                for hash in hashes:
+                    if nodes.get(key, {}).get(hash, {}) == {}:
+                        nodes.setdefault(key, {})
+                        nodes[key][hash] = extracted[key][hash]
+                    else:
+                        for rel_uuid, rel in extracted[key][hash]["relationships"].items():
+                            nodes[key][hash]["relationships"][rel_uuid] = rel
+            return nodes
+
+        def merge(nodes1, nodes2):
```

Comment:
can you add documentation about how resolution is currently working?

Reply:
added documentation in ResolveEntities. Note, we will probably move this code to a seperate transform later on.